### DISCUSSION TITLE
NodeJS Project - Food24x7

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#102] [Tasks]
![image](https://user-images.githubusercontent.com/60097560/188258115-02cc3084-7dcd-4c1a-bc6d-98504d79407f.png)
![image](https://user-images.githubusercontent.com/60097560/188258132-032b0776-d1a0-4c06-8982-343c4eedc06b.png)
![image](https://user-images.githubusercontent.com/60097560/188258140-5a0bd730-01dc-47e0-80f2-9537c2901663.png)

**Comment by shreysingh8838** [comment]
const express =  require("express");
const mongoose = require("mongoose");
const jwt = require("jsonwebtoken");

const user = require("./UserModel");

const app = express();
app.use(express.json());

app.post("/api/v1/login", async(req, res, next)=>{
    let{ username, password } = req.body;

    let existinguser;
    try{
        existinguser = user.findOne({username: username});
    }catch{
        const error = new Error("Something went wrong1");
        return next(error);
    }

    if(!existinguser || existinguser.password != password){
        const error = new Error("Wrong credentials entered");
        return next(error);
    }

    let token;

    try{
        token = jwt.sign(
            { userid : existinguser.id}, "mysecret", {expiresIn: "1h"}
        );
    } catch(err){
        console.log(err);
        const error = new Error("Something went wrong");
        return next(error); 
    }

    res.status(200).json({
        success : true,
        data: {
            userid,
            token,
        },
    })
});


mongoose
  .connect("mongodb://localhost:27017/testDB")
  .then(() => {
    app.listen("3000", () => {
      console.log("Server is listening on port 3000");
    });
  })
  .catch((err) => {
    console.log("Error Occurred");
  });``

**Reply by vivekvikigupta** [reply]
Hi Shreyas,
Can we use "await" to get resolved promise else it might give error as we are using async function.
i.e. : existinguser = await user.findOne({username: username});

**Reply by shreysingh8838** [reply]
Yeah I am really sorry i have not updated the code actually, 

JWT LOGIN
` app.post("/api/v1/login", async(req, res, next)=>{
    let{ username, password } = req.body;

    let existinguser;
    try{
        existinguser = await user.findOne({"username": username});
        console.log(existinguser);
    }catch{
        const error = new Error("Something went wrong1");
        return next(error);
    }

    if(!existinguser || existinguser.password != password){
        const error = new Error("Wrong credentials entered");
        console.log(`${existinguser.password} & ${password}`)
        return next(error);
    }
    let token;
    try{
        token = jwt.sign(
            { userid : existinguser.id}, "mysecret", {expiresIn: "1h"}
        );
    }catch(err){
        console.log(err);
        const error = new Error("Something went wrong");
        return next(error); 
    }

    res.status(200).json({
        success : true,
        data: {
            userid : existinguser._id,
            token,
        },
    })
});
`

JWT SIGNUP
`app.post("/api/v1/signup", async (req, res, next) => {
    const { username, password, email } = req.body;
    const newUser = new user({
      username,
      email,
      password,
    });
    console.log(newUser);
   
    try {
      await newUser.save();
    } catch {
      const error = new Error("Error! Something went wrong.");
      return next(error);
    }
    let token;
    try {
      token = jwt.sign(
        { userId: newUser.id, email: newUser.email },
        "mysecretkey",
        { expiresIn: "1h" }
      );
    } catch (err) {
      const error = new Error("Error! Something went wrong.");
      return next(error);
    }
    res.status(201)
      .json({
        success: true,
        data: { status : "User created successfully" },
        token
      });
  });
`
